pycalphad/tests/test_equilibrium.py

```diff
@@ -53,6 +53,39 @@ def test_eq_binary():
     eqx = equilibrium(ALFE_DBF, comps, my_phases, conds, verbose=True)
     assert_allclose(eqx.GM.values.flat[0], -9.608807e4)
 
+
+def test_phase_records_passed_to_equilibrium():
+    "Pre-built phase records can be passed to equilibrium."
+    my_phases = ['LIQUID', 'FCC_A1', 'HCP_A3', 'AL5FE2', 'AL2FE', 'AL13FE4', 'AL5FE4']
+    comps = ['AL', 'FE', 'VA']
+    conds = {v.T: 1400, v.P: 101325, v.N: 1.0, v.X('AL'): 0.55}
+
+    models = instantiate_models(ALFE_DBF, comps, my_phases)
+    phase_records = build_phase_records(ALFE_DBF, comps, my_phases, conds, models)
+
+    # Without models passed
+    eqx = equilibrium(ALFE_DBF, comps, my_phases, conds, verbose=True, phase_records=phase_records)
```

Comment:
I don't think the models being rebuilt when phase records are passed in is good behavior, from a least-surprise perspective. This could lead to surprising, non-intuitive results for custom models. I think either (a) `phase_records` being passed in should require `models` being passed in, or (b) somehow the `models` should be inferred from the `phase_records` without being rebuilt.

Reply:
I agree. I'm leaning towards (a) because I don't want to enforce coupling between phase records and models. 

One complication is that the `model` argument can be unpacked via `unpack_kwarg` and it can have many forms as input, so it may be tricky to validate.  Some options of how to handle it in the case that `phase_records` is passed:

1. Do not `instantiate_models`, but make a best effort check that `all(isinstance(mod, Model) for mod in model.values())`. This alone wouldn't catch the case where the `model` argument only overrides a subset of active phases (and `instantiate_models` would typically build the rest), but we could address that in a similar way.
2. Do not instantiate models. Don't do any validation and just pass whatever the `model` argument was to `build_phase_records`, letting `build_phase_records` and `build_callables` raise appropriately.

Reply:
I think the first option is probably reasonable and likely to give good error messages.

Reply:
Here's the commit that adds code to handle this behavior in the spirit of the first option: https://github.com/pycalphad/pycalphad/pull/361/commits/a13cbcc6a3453ba415a1fee3ea72b07c137fae70

It's safer, but doesn't really protect users from providing incompatible model instances. The user is still responsible for making sure their argument to `model` is actually correct. Model objects really only serve two purposes in the `equilibrium` code path:

1. building phase records
2. getting the size of the internal degrees of freedom and the indices of pure vacancy endmembers for point sampling

if we could somehow do the second one without a Model instance, we could make it so providing a phase record or model are mutually exclusive per phase.

Reply:
I think this approach is fine for the time being.